### PR TITLE
Do not explicitly call gc.collect in `CompImageHDU.compressed_data`

### PR DESCRIPTION
**Pull request by Cadair**
I have a nice pathological benchmark where I load 150 files and read all the data:

Before:

```
$ python run_timeit.py dkist_visp_tiled_post
Using DKIST dataset (BEJZP) with shape: (4, 18, 150, 996, 2545)
FITS tile shape is (1, 256, 256) (numpy order).
Ran dkist_visp_tiled_post 5 times, average time 27.755185409799743 s
```

After:

```
$ python run_timeit.py dkist_visp_tiled_post
Using DKIST dataset (BEJZP) with shape: (4, 18, 150, 996, 2545)
FITS tile shape is (1, 256, 256) (numpy order).
Ran dkist_visp_tiled_post 5 times, average time 4.998688907800533 s
```

We already discuss the reason why this line exists pretty comprehensively in the docs here: https://docs.astropy.org/en/stable/io/fits/appendix/faq.html#i-am-opening-many-fits-files-in-a-loop-and-getting-oserror-too-many-open-files

I am not really sure I see the benefit of this still being here?

### CODE REVIEW

**Comment by astrofrog** [comment]
I agree that users who run into issues should simply call gc.collect themselves - we shouldn't call it by default as this can have a large performance penalty.

**Comment by pllim** [comment]
But people been relying on this indirectly for so long, so now suddenly they have to manually call `gc.collect()` to get back the same behavior?

**Comment by astrofrog** [comment]
@pllim - I'm not sure it would actually change anything for most users, we shouldn't actually have to call garbage collection explicitly and it's surprising that the code even includes this. It might have been added as a 'just in case' but the main use case it might have been intended for is opening many files in a small amount of time which is what @Cadair is doing but it seems to be hindering not helping.

**Comment by pllim** [comment]
Is asking over at astropy-dev mailing list necessary, just in case?

**Comment by Cadair** [comment]
I can't imagine that even if anyone is relying on this they know it exists, so asking wont help.

**Comment by pllim** [comment]
Should at least add a change log then?

**Comment by Cadair** [comment]
> Should at least add a change log then?

Sure. I was waiting to see if people a) rejected it outright and b) that all the tests passed :grinning:

**Comment by Cadair** [comment]
For what it's worth, I just loaded 10800 FITS files with memmap enabled without this line and I didn't hit any issues. I am not sure when dask would have dropped the references to them, so I don't know how many were open at once, but it certainly wasn't keeping them around forever.

**Comment by saimn** [comment]
I agree we should not call `gc.collect` since it will have a performance penalty when processing a lot of files. 
For reference, this was added in #3283 / cf51894fc6ed3c50c198b1e6ab2980b59916b73b.

> For what it's worth, I just loaded 10800 FITS files with memmap enabled without this line and I didn't hit any issues

Interesting, though problems may arise only in specific configurations, e.g.  did you take care of not keeping references to `.data` ? What the value of `ulimit -n` on your system ? 

I guess the other case to check, where people could be impacted by this change, is reading one or a few big files, keeping / not keeping references to `.data`, to see if memory if released or not.

**Comment by Cadair** [comment]
> did you take care of not keeping references to .data ? 

I didn't but dask might have done.

> What the value of ulimit -n on your system ? 

```
$ ulimit -n
1024
```

I am not suggesting this is a comprehensive test.

File: docs/changes/io.fits/14576.feature.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Do not call ``gc.collect()`` in ``CompImageHDU.compressed_data`` as it has
```
Comment (astrofrog) [inline docs/changes/io.fits/14576.feature.rst:1]: Specifically in the deleter right? Maybe rephrase to say when closing a ConpImageHDU since that is when people are most likely to call this indirectly?

**Comment by Cadair** [comment]
@astrofrog @saimn I think this is good to go?

**Comment by saimn** [comment]
I did a few quick tests and I'm confused, it seems we have a memory leak with the new code ?
https://github.com/saimn/misc-astro/blob/master/astropy/FITS%20compression.ipynb
Note, for dev I get the same memory usage with `gc.collect` or without.
Also the amount of memory that leaks seems to vary with the tiling, it's bigger with the default tiling.

**Comment by saimn** [comment]
Running again some tests after #14649, I don't see a difference with or without `gc.collect()`, so sounds good to me.

**Review by saimn — APPROVED** [review]
Thanks @Cadair !
